### DISCUSSION TITLE
New "switch" validation rule

### DISCUSSION THREAD

**Discussion by dbfun** [laravel/framework#48004] [Show and tell]
Sometimes I need to turn off form validation rules.
This can be done by adding several "if". However, this adds a lot of code.
I am using an another way.

See:

```php
[
    'date' => 'switch|required|date_format:Y-m-d', // "Switch" works here.
    'email' => 'required|string|email', // "Switch" doesn't work here.
];
```

If `withValidation=true` is passed in a query, then the validation rules are used.

If `withValidation` is not present, all validation rules for this field are ignored.

The "switch" is controlled via the middleware:

```php
Route::group([
    'middleware' => ValidationSwitch::class,
], function () {
    Route::post('/', [FormController::class, 'handle'])->name('form');
});
```

In our company, we enable validation for partners API and disable it for the site API.

If you like the idea, I'll make a pull request.

**Comment by henzeb** [comment]
Not a big fan of the idea. as the validation is a way to prevent garbage data. Why would you turn it off for your site? it's not likely your site will send in something else than a date, as per your example. Or give me a real life use case that makes more sense than this.

also, why handling this with a query parameter? I could be a partner. I could just want to enter a datetime there without going through the hassle of modifying it's format and then ommit 'withValidation'...

**Reply by dbfun** [reply]
> Why would you turn it off for your site?

At the very beginning, we did not have validation on the back, but there was validation on the front. The form was received on the back and then sent to another system that has own validation. I work in a bank and we needed to reduce form loss due to validation. Back works as a proxy. There should be three places for validation: site (JS), API, end system. We did not begin to do validation on the API, this is the decision of the management.

This suited everyone. Then we start working with partners. They started sending invalid data. It is very difficult to validate on the back for 50 forms and not break anything for existing forms. Because there are a lot of complex validation rules. And there is no easy way to copy validation from JS to PHP.

Therefore, we skip the validation of many fields from the site, but give partners the opportunity to validate the data.

> also, why handling this with a query parameter? I could be a partner. I could just want to enter a datetime there without going through the hassle of modifying it's format and then ommit 'withValidation'...

We give the partner a choice - either to validate the form or not. If the end banking system receives invalid data, the application is rejected and the partners do not receive a reward. Partners can either do validation on their front or use our validation if they pass a parameter.

I agree that this is a strange decision. But the reasons why this is done is that it takes a lot of time to synchronize the validation rules, and the testing team is very slow.

But I think it might be helpful. The "switch" can be enabled by any condition, it does not have to be a parameter in the request.

Benefits that our team personally received

1. The first reason is less code.

Before:

```php
$rules = [
    'phone' => 'bail|required|digits:11|starts_with:7',
];

if(some_conditions()) {
    $rules['code'] = 'required|string|digits:4';
}
```


After:

```php
$rules = [
    'phone' => 'bail|required|digits:11|starts_with:7',
    'code' => 'switch|required|string|digits:4',
];
```

2. The second reason is that validation rules can be stored in a database. We now have about 50 forms. The largest form has 50 validation fields. Without this rule, it is difficult to save the conditions for enabling validation for individual fields in the database.

**Reply by sohelrana820** [reply]
It seems you have a very customized requirement from your partner. But in general, I think it would not good idea to integrate this laravel validation section.

**Reply by sohelrana820** [reply]
Can you do something by using "nullable" rule of laravel

**Comment by hypnodev** [comment]
Is this the case to use [sometimes](https://laravel.com/docs/10.x/validation#validating-when-present)?

**Reply by dbfun** [reply]
No. The enable condition is controlled from the code. We have this parameter in the request. But you can make any other condition at the middleware level or elsewhere.

**Comment by Angel5a** [comment]
Something like `exclude_if` but preserving field in validated output?

I see use-case for this in terms of premium users, but in general, not only switching off. Like, if you a premium, you have a max of 20, otherwise 10. As it sounds, so it is writen:
```php
$rules = [
    'phone' => 'bail|required|digits:11|starts_with:7',
    'code' => some_conditions() ? 'required|string|digits:4' : 'sometimes',
];
```
